### PR TITLE
Add support for updating AWS routing table entries

### PR DESCRIPTION
**Pull request by geofffranks**
The `advertised_instance_routes` cloud property on resource pools/vm
types has been added to allow manifests to connect routes on AWS
Routing Tables with specific instances, if you need to route traffic
through a BOSH deployed VM (like in the case of a site-to-site VPN).

If using this, it is recommended to only have a single instance of a
single job of that resource pool, to avoid confusion. If there are
multiple instances/jobs in the resource pool, the last one processed
will be the node assigned to the route.

Datastructure of `advertised_instance_routes` is as follows:

```
cloud_properties:
  advertised_instance_routes:
  - table_id:    rt-abcdef123
    destination: 10.0.0.0/16
  - table_id:    rt 12345abcd
    destination: 10.0.0.0/16
```

This allows the manifest to specify that multiple routing tables
are updated to send the destination route(s) to it. Once removed
from the manifest, the routes will remain, but once the VM is recreated,
the route will become a black hole in the AWS routing table, and cease
to work.

### CODE REVIEW

File: src/bosh_aws_cpi/lib/cloud/aws/instance.rb [hunk 1]
```diff
@@ -83,6 +83,30 @@ def exists?
...
+      if route_definitions.length > 0
+        @logger.debug("Associating instance with destinations in the routing tables")
+        tables = @aws_instance.vpc.route_tables
+        route_definitions.each do |definition|
+          @logger.debug("Finding routing table '#{definition["table_id"]}'")
+          table = tables[definition["table_id"]]
+          @logger.debug("Sending traffic for '#{definition["destination"]}' to '#{@aws_instance.id}' in '#{definition["table_id"]}'")
+
+          existing = false
+          table.routes.each do |route|
```
Comment (ljfranklin) [inline src/bosh_aws_cpi/lib/cloud/aws/instance.rb]: Docs say `routes` is a normal array, so could probably shorten this to: `existing = table.routes.any ...`

**Comment by dpb587** [comment]
Oh, that's an interesting idea. Hadn't considered this sort of thing as a CPI property before, just as a co-located release sort of thing. Wonder what CPI team thinks of this approach.

**Comment by cppforlife** [comment]
@dpb587 api already talks to the iaas so might as well have it here instead of connecting from somewhere else. it's logged, diffed via cloud properties, audited through events, etc.

**Comment by ljfranklin** [comment]
@geofffranks neat feature! Looks good, my only feedback is seems a little messy to not delete the route when we delete the VM. Do you see any issues with calling `delete_route` during `delete_vm` for any routes that have `route.instance == instance_to_delete`, or is there some reason why a black-hole route is desirable?

**Comment by dpb587** [comment]
Cool - I'll look forward to using it!

**Comment by cppforlife** [comment]
@ljfranklin black holes are not bad. just means not configured. i think we should make as little api calls as possible. cc @geofffranks

**Comment by geofffranks** [comment]
I can add the delete feature, and clean up the check for existing routes.

Thanks for the quick feedback!

**Comment by geofffranks** [comment]
Just saw @cppforlife 's note about maybe not deleting. Let me know which way you want to go.

**Comment by ljfranklin** [comment]
@cppforlife @geofffranks If I'm relying on BOSH to create a route for me, my expectation would be for BOSH to also remove that route when that VM goes away. I could imagine an initial setup with a fallback route `0.0.0.0` that goes to a Gateway. You run `bosh deploy vpn-box.yml` with `advertised_instance_routes`, all traffic to the given CIDR magically starts getting routed through the VPN box. You later run `bosh delete deployment vpn-box.yml` and everything magically starts going back through the Gateway. Pretty slick IMO.

**Comment by ljfranklin** [comment]
After a convo on slack, the consensus is to leave the black hole routes as we don't have access to the `cloud_properties` during the [delete_vm](http://bosh.io/docs/cpi-api-v1.html#delete-vm) call. Seems sketchy to delete routes that may have been created out-of-band. Story to merge PR is [here](https://www.pivotaltracker.com/story/show/129380985).

**Comment by cppforlife** [comment]
@geofffranks Build failed on integration tests due to rubygems flakiness. Restarted.
